File: DataLakeOperations.cs
using System;
using System.IO;
using System.Threading.Tasks;
using Microsoft.AspNetCore.Mvc;
using Microsoft.Azure.WebJobs;
using Microsoft.Azure.WebJobs.Extensions.Http;
using Microsoft.AspNetCore.Http;
using Microsoft.Extensions.Logging;
using Azure.Storage.Files.DataLake;
using Azure.Identity;
using System.Text;
using System.Collections.Generic;
using System.Linq;

namespace BasicHealthcareFunctions
{
    public static class DataLakeOperations
    {
        [FunctionName("UploadToDataLake")]
        public static async Task<IActionResult> UploadFile(
            [HttpTrigger(AuthorizationLevel.Function, "post", Route = "datalake/upload/{container}/{fileName}")] HttpRequest req,
            string container,
            string fileName,
            ILogger log)
        {
            log.LogInformation($"Upload request for file: {fileName} to container: {container}");

            try
            {
                var storageAccountName = Environment.GetEnvironmentVariable("DATA_LAKE_STORAGE_ACCOUNT");
                if (string.IsNullOrEmpty(storageAccountName))
                {
                    return new BadRequestObjectResult("Storage account configuration not found.");
                }

                // Validate container name
                if (!IsValidContainer(container))
                {
                    return new BadRequestObjectResult("Invalid container name. Valid containers: raw, processed, curated");
                }

                // Use managed identity to authenticate
                var credential = new DefaultAzureCredential();
                var dataLakeServiceClient = new DataLakeServiceClient(
                    new Uri($"https://{storageAccountName}.dfs.core.windows.net"), 
                    credential);

                var fileSystemClient = dataLakeServiceClient.GetFileSystemClient(container);
                var fileClient = fileSystemClient.GetFileClient(fileName);

                // Read the request body
                using var reader = new StreamReader(req.Body);
                var content = await reader.ReadToEndAsync();

                // Upload the file
                using var stream = new MemoryStream(Encoding.UTF8.GetBytes(content));
                await fileClient.UploadAsync(stream, overwrite: true);

                log.LogInformation($"Successfully uploaded file: {fileName} to container: {container}");

                return new OkObjectResult(new
                {
                    Message = "File uploaded successfully",
                    Container = container,
                    FileName = fileName,
                    Timestamp = DateTime.UtcNow
                });
            }
            catch (Exception ex)
            {
                log.LogError(ex, $"Failed to upload file: {fileName}");
                return new ObjectResult(new { Error = ex.Message }) { StatusCode = 500 };
            }
        }

        [FunctionName("ListDataLakeFiles")]
        public static async Task<IActionResult> ListFiles(
            [HttpTrigger(AuthorizationLevel.Function, "get", Route = "datalake/list/{container}")] HttpRequest req,
            string container,
            ILogger log)
        {
            log.LogInformation($"List files request for container: {container}");

            try
            {
                var storageAccountName = Environment.GetEnvironmentVariable("DATA_LAKE_STORAGE_ACCOUNT");
                if (string.IsNullOrEmpty(storageAccountName))
                {
                    return new BadRequestObjectResult("Storage account configuration not found.");
                }

                // Validate container name
                if (!IsValidContainer(container))
                {
                    return new BadRequestObjectResult("Invalid container name. Valid containers: raw, processed, curated");
                }

                // Use managed identity to authenticate
                var credential = new DefaultAzureCredential();
                var dataLakeServiceClient = new DataLakeServiceClient(
                    new Uri($"https://{storageAccountName}.dfs.core.windows.net"), 
                    credential);

                var fileSystemClient = dataLakeServiceClient.GetFileSystemClient(container);
                var files = new List<object>();

                await foreach (var pathItem in fileSystemClient.GetPathsAsync())
                {
                    files.Add(new
                    {
                        Name = pathItem.Name,
                        IsDirectory = pathItem.IsDirectory,
                        LastModified = pathItem.LastModified,
                        ContentLength = pathItem.ContentLength
                    });
                }

                log.LogInformation($"Successfully listed {files.Count} items in container: {container}");

                return new OkObjectResult(new
                {
                    Container = container,
                    Files = files,
                    Count = files.Count,
                    Timestamp = DateTime.UtcNow
                });
            }
            catch (Exception ex)
            {
                log.LogError(ex, $"Failed to list files in container: {container}");
                return new ObjectResult(new { Error = ex.Message }) { StatusCode = 500 };
            }
        }

        [FunctionName("DownloadFromDataLake")]
        public static async Task<IActionResult> DownloadFile(
            [HttpTrigger(AuthorizationLevel.Function, "get", Route = "datalake/download/{container}/{fileName}")] HttpRequest req,
            string container,
            string fileName,
            ILogger log)
        {
            log.LogInformation($"Download request for file: {fileName} from container: {container}");

            try
            {
                var storageAccountName = Environment.GetEnvironmentVariable("DATA_LAKE_STORAGE_ACCOUNT");
                if (string.IsNullOrEmpty(storageAccountName))
                {
                    return new BadRequestObjectResult("Storage account configuration not found.");
                }

                // Validate container name
                if (!IsValidContainer(container))
                {
                    return new BadRequestObjectResult("Invalid container name. Valid containers: raw, processed, curated");
                }

                // Use managed identity to authenticate
                var credential = new DefaultAzureCredential();
                var dataLakeServiceClient = new DataLakeServiceClient(
                    new Uri($"https://{storageAccountName}.dfs.core.windows.net"), 
                    credential);

                var fileSystemClient = dataLakeServiceClient.GetFileSystemClient(container);
                var fileClient = fileSystemClient.GetFileClient(fileName);

                // Check if file exists
                var exists = await fileClient.ExistsAsync();
                if (!exists.Value)
                {
                    return new NotFoundObjectResult($"File {fileName} not found in container {container}");
                }

                // Download the file
                var response = await fileClient.ReadAsync();
                using var reader = new StreamReader(response.Value.Content);
                var content = await reader.ReadToEndAsync();

                log.LogInformation($"Successfully downloaded file: {fileName} from container: {container}");

                return new OkObjectResult(new
                {
                    Container = container,
                    FileName = fileName,
                    Content = content,
                    Timestamp = DateTime.UtcNow
                });
            }
            catch (Exception ex)
            {
                log.LogError(ex, $"Failed to download file: {fileName}");
                return new ObjectResult(new { Error = ex.Message }) { StatusCode = 500 };
            }
        }

        private static bool IsValidContainer(string container)
        {
            var validContainers = new[] { "raw", "processed", "curated" };
            return validContainers.Contains(container.ToLowerInvariant());
        }
    }
}

File: datalake_operations.py
import azure.functions as func
import logging
import json
import os
from azure.storage.blob import BlobServiceClient
from azure.identity import DefaultAzureCredential
from function_app import app

# Add data lake operations to the same function app
@app.route(route="datalake/upload", methods=["POST"])
def upload_to_datalake(req: func.HttpRequest) -> func.HttpResponse:
    """
    Upload files to the data lake storage account
    """
    logging.info('Python HTTP trigger function for data lake upload.')
    
    try:
        # Get the storage account connection from environment variables
        storage_account_url = os.environ.get('STORAGE_ACCOUNT_URL')
        
        if not storage_account_url:
            return func.HttpResponse(
                json.dumps({"error": "Storage account URL not configured"}),
                status_code=500,
                headers={"Content-Type": "application/json"}
            )
        
        # Use managed identity for authentication
        credential = DefaultAzureCredential()
        blob_service_client = BlobServiceClient(account_url=storage_account_url, credential=credential)
        
        # Get request body
        req_body = req.get_json()
        if not req_body:
            return func.HttpResponse(
                json.dumps({"error": "Request body is required"}),
                status_code=400,
                headers={"Content-Type": "application/json"}
            )
        
        container_name = req_body.get('container', 'healthcare-data')
        blob_name = req_body.get('blob_name')
        content = req_body.get('content')
        
        if not blob_name or not content:
            return func.HttpResponse(
                json.dumps({"error": "blob_name and content are required"}),
                status_code=400,
                headers={"Content-Type": "application/json"}
            )
        
        # Upload to blob storage
        blob_client = blob_service_client.get_blob_client(container=container_name, blob=blob_name)
        blob_client.upload_blob(content, overwrite=True)
        
        response = {
            "status": "success",
            "message": f"File {blob_name} uploaded successfully to {container_name}",
            "blob_url": blob_client.url
        }
        
        return func.HttpResponse(
            json.dumps(response),
            status_code=200,
            headers={"Content-Type": "application/json"}
        )
        
    except Exception as e:
        logging.error(f"Upload failed: {str(e)}")
        return func.HttpResponse(
            json.dumps({"error": f"Upload failed: {str(e)}"}),
            status_code=500,
            headers={"Content-Type": "application/json"}
        )

@app.route(route="datalake/list", methods=["GET"])
def list_datalake_files(req: func.HttpRequest) -> func.HttpResponse:
    """
    List files in the data lake storage account
    """
    logging.info('Python HTTP trigger function for listing data lake files.')
    
    try:
        # Get the storage account connection from environment variables
        storage_account_url = os.environ.get('STORAGE_ACCOUNT_URL')
        
        if not storage_account_url:
            return func.HttpResponse(
                json.dumps({"error": "Storage account URL not configured"}),
                status_code=500,
                headers={"Content-Type": "application/json"}
            )
        
        # Use managed identity for authentication
        credential = DefaultAzureCredential()
        blob_service_client = BlobServiceClient(account_url=storage_account_url, credential=credential)
        
        container_name = req.params.get('container', 'healthcare-data')
        
        # List blobs in container
        container_client = blob_service_client.get_container_client(container_name)
        blobs = []
        
        for blob in container_client.list_blobs():
            blobs.append({
                "name": blob.name,
                "size": blob.size,
                "last_modified": blob.last_modified.isoformat() if blob.last_modified else None,
                "content_type": blob.content_settings.content_type if blob.content_settings else None
            })
        
        response = {
            "status": "success",
            "container": container_name,
            "blob_count": len(blobs),
            "blobs": blobs
        }
        
        return func.HttpResponse(
            json.dumps(response),
            status_code=200,
            headers={"Content-Type": "application/json"}
        )
        
    except Exception as e:
        logging.error(f"List operation failed: {str(e)}")
        return func.HttpResponse(
            json.dumps({"error": f"List operation failed: {str(e)}"}),
            status_code=500,
            headers={"Content-Type": "application/json"}
        )
